### DISCUSSION TITLE
How does the volume server determine whether to include the "mime" property in the upload response?

### DISCUSSION THREAD

**Discussion by m59peacemaker** [seaweedfs/seaweedfs#4127] [General]
In the README `curl` examples, there is no "mime" property on the response from the volume server when uploading files. However, The `mime` property is included in the response for some file types. In my testing with the request including the content type in the multipart form data, the volume server response includes the mime property for all upload requests except for most image types. When I upload a `image/bmp` file, the response includes `mime`, but when I upload `image/jpeg`, `image/png`, or `image/webp` files, the response does not include `mime`. It seems that for some file types, the provided content type is included in the response, and for others, it is not. Is this intentional? If so, what determines whether the `mime` property will be included in the response?

**Comment by m59peacemaker** [comment]
I've been reading the source code trying to figure this out. It does seem like a bug to me because it looks like `mime` should always be present in this response when it is supplied with the uploaded file, especially because the volume server always sends the correct content-type when I retrieve the files, and jpg rotation is working in all of these cases:
```sh
curl -v -F 'file=@my-image.jpg' "http://${host}/${id}"
# { name: 'my-image.jpg', size, eTag }

curl -v -F 'file=@my-image.jpg;type=image/jpeg' "${host}/${id}"
# { name: 'my-image.jpg, size, eTag }

curl -v -F 'file=@my-image;type=image/jpeg' "http://${host}/${id}"
# { name: 'my-image', size, eTag, mime: 'image/jpeg' }
```

It is interesting that the last request has `mime` in the response while the others don't. So it appears to be something about the file extension that makes the difference, but I don't see it in the code yet.

My greatest suspicion is this:
https://github.com/seaweedfs/seaweedfs/blob/c42b95c596f762dcca2bc9c7e7a918ab8ca8b206/weed/util/compression.go#L111-L150

It examines extensions before mime types, and handles bmp differently than the other image types. I know `bmp` uploads to include `mime` in the response, while other images don't. So I'm looking for a condition using `IsCompressableFileType` that leads to `mime` being lost. But having searched the codebase for `IsCompressableFileType` and looking at all usages, I don't see anything that affects `Mime` or `MimeType`. So I'm possibly hitting a wall now.

**Comment by chrislusf** [comment]
There are multiple ways to add a file, http, s3, fuse mount, etc. Each is a bit different.

Here are the ways for server to find out the mime with best effort:
1. http mime type header
2. mime type detection on the first 512 bytes
3. add mime type by file name extension during download process.

The 2 and 3 options are depending on the machine's local mime type settings.

**Comment by m59peacemaker** [comment]
In all cases, my requests explicitly include the content-type for the file in the multipart form data (not to be confused with the content-type header, which is multipart/form-data). In case I have not communicated well, this is a light reproduction of my issue:

```sh
# get the fid
curl "${host}/dir/assign"
# jpeg but no extension on the filename, explicitly include image/jpeg content-type
curl -F 'file=@my-file;type=image/jpeg' "http://${host}/${fid}"
# response has `mime` property
# { name: 'my-file', mime: 'image/jpeg', size, eTag }

# get the fid
curl "${host}/dir/assign"
# jpeg, with extension on the filename, explicitly include image/jpeg content-type
curl -F 'file=@my-file.jpg;type=image/jpeg' "http://${host}/${fid}"
# response does NOT have `mime` property
# { name: 'my-file.jpg', size, eTag }
```

Most file types do not work this way. For example:

```sh
# get the fid
curl "${host}/dir/assign"
curl -F 'file=@my-file.whatever;type=lol/nonsense' "http://${host}/${fid}"
# response has `mime` property
# { name: 'my-file.whatever', mime: 'lol/nonsense', size, eTag }
```

All file types I provide except some image types will have `mime` in the response from the volume server. I'm trying to understand why these image types don't work the same way.

**Comment by m59peacemaker** [comment]
@chrislusf I found the two lines of code that make the difference.

https://github.com/seaweedfs/seaweedfs/blob/master/weed/storage/needle/needle_parse_upload.go#L197-L198

```go
if contentType != "" && contentType != "application/octet-stream" && mtype != contentType {
  pu.MimeType = contentType // only return mime type if not deducible
```

It's because `mtype == "image/jpeg"` and `contentType == "image/jpeg", so the above condition is `false`, and so `pu.MimeType` is not set. The comment seems to confirm the intention: "only return mime type if not deducible". I would like to understand more about this.
